validator/sawtooth_validator/journal/publisher.py

```diff
@@ -833,6 +833,7 @@ def update(self, still_pending, last_sent):
         last_index = self._batches.index(last_sent)
         unsent = self._batches[last_index + 1:]
         self._batches = still_pending + unsent
+        self._ids = set(b.header_signature for b in self._batches)
```

Comment:
Why not just replace `self._ids` with a generator?  Seems like it would be better to use `self._batches` as the source of truth, and the ids are just a view.  Secondly, making it a set where as self._batches is a list seems unnecessary or is missing a check that the should be same length - any discrepancy implies there is a duplicate batch in there.

Reply:
The main benefit of using a set like this is improved lookup times. You are correct to point out that it does not correctly handle duplicate batches though, so I will add a check around that.